app-cpp/appPtr.hpp

```diff
@@ -0,0 +1,481 @@
+        {
+            reset();
+
+            if (inCopy)
+            {
+                inCopy.mControl->IncrementStrong();
+                mData = inCopy.mData;
+                mControl = inCopy.mControl;
+            }
+        }
+        return *this;
+    }
+
+    T* operator->()
+    {
+        return get();
+    }
+
+    T& operator*()
+    {
+        return *get();
+    }
+
+    operator bool() const
+    {
+        return (mData != nullptr);
+    }
+
+    T* get()
+    {
+        return mData;
+    }
+
+private:
+    class ControlBlock
+    {
+    public:
+        ControlBlock(T* inData) :  
+            mData{inData}
+        {
+            // Do nothing
+        }
+
+        // RO5 Methods added
+        ~ControlBlock() = default;
+
+        // Copying should not be done, because we use a std::mutex
+        ControlBlock(const ControlBlock& inCopy) = delete;
+        ControlBlock& operator=(const ControlBlock& inCopy) = delete;
+
+        std::pair<std::size_t, std::size_t> IncrementStrong()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            const bool hasDataBeenDeleted = (mStrongCount <= 0);
+            if (!hasDataBeenDeleted)
+            {
+                ++mStrongCount;
+            }
+            return {mStrongCount, mWeakCount};
+        }
+
+        std::pair<std::size_t, std::size_t> DecrementStrong()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            // StrongCount is not allowed to change once it reaches 0 and must be deleted
+            assert((mStrongCount > 0) && "Cannot decrement below zero");
+            --mStrongCount;
+            return {mStrongCount, mWeakCount};
+        }
+
+        std::pair<std::size_t, std::size_t> IncrementWeak()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            ++mWeakCount;
+            return {mStrongCount, mWeakCount};
+        }
+
+        std::pair<std::size_t, std::size_t> DecrementWeak()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            // StrongCount is not allowed to change once it reaches 0 and must be deleted
+            assert((mWeakCount > 0) && "Cannot decrement below zero");
+            --mWeakCount;
+            return {mStrongCount, mWeakCount};
+        }
+
+        T* get()
+        {
+            return mData;
+        }
+
+    private:
+        std::size_t mStrongCount{1}; // Default constructs with initial refcount of 1
+        std::size_t mWeakCount{0};
+        std::mutex mMutex{};
+        T* mData{}; // Used by weak pointers to obtain a new shared pointer
+    }; // class ControlBlock
+
+private:
+    template <typename T>
+    friend class weak_ptr;
+
+    shared_ptr(ControlBlock* inBlock)
+    {
+        if (inBlock != nullptr)
+        {
+            inBlock->IncrementStrong();
+            mData = inBlock->get();
+            mControl = inBlock;
+        }
+    }
+
+private:
+    T* mData{};
+    ControlBlock* mControl{};
```

Comment:
Since these 2 types are pointers, and pointers are an intrinsic type (meaning: already-defined as part of the language, and not some custom type of your own creation)... I would make the default initialization explicit.
```
    T* mData{nullptr};
    ControlBlock* mControl{nullptr};
```

I would do this for any intrinsic type (for instance: `int`, or `bool`, etc)